### PR TITLE
[DP-811] POC: Porting canton's approach for the security evidencing

### PR DESCRIPTION
**Pull request by skisel-da**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox-on-x/BUILD.bazel [hunk 1]
```diff
@@ -194,11 +196,10 @@ da_scala_test_suite(
     ],
 )
 
-da_scala_test_suite(
+da_scala_library(
```
Comment (skisel-da) [inline ledger/sandbox-on-x/BUILD.bazel:199]: Here we have a problem as we want to access test_suite modules in another module so we could inspect all test suites within the classpath.

File: ledger/ledger-security-test-tool/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        "@maven//:com_github_pathikrit_better_files",
+        "@maven//:com_lihaoyi_sourcecode",
+
+    ],
+    tags = ["maven_coordinates=com.daml:ledger-security-test-tool:__VERSION__"],
+    visibility = [
+        "//visibility:public",
+    ],
+    deps = [
+        "//ledger/ledger-security-test-lib",
```
Comment (tudor-da) [inline ledger/ledger-security-test-tool/BUILD.bazel:27]: Can we have them in the same Bazel package? They are both pertaining to security evidencing, right?

Comment (skisel-da) [reply ledger/ledger-security-test-tool/BUILD.bazel:27]: I think most of the test suite code will be then dependent on `ledger-security-test-lib`, but they dont need to be dependent on the tool code, but for simplification we could do that, yes...

File: ledger/ledger-security-test-lib/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,32 @@
+# Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+load("//bazel_tools:scala.bzl", "da_scala_library")
+
+da_scala_library(
+    name = "ledger-security-test-lib",
```
Comment (tudor-da) [inline ledger/ledger-security-test-lib/BUILD.bazel:7]: I would name the entire package `ledger-security-evidence` or similar and have both targets defined here in it

File: ledger/sandbox-on-x/BUILD.bazel (hunk 1 above)
Comment (tudor-da) [reply ledger/sandbox-on-x/BUILD.bazel:199]: but now these tests are not run. this is WIP, right?

**Comment by tudor-da** [comment]
I think the direction is good! Left some comments, while not doing a deep-dive as I understand this is a WIP.

File: ledger/sandbox-on-x/BUILD.bazel (hunk 1 above)
Comment (skisel-da) [reply ledger/sandbox-on-x/BUILD.bazel:199]: they're not working, yes.

File: ledger/ledger-security-test-tool/BUILD.bazel (hunk 2 above)
Comment (tudor-da) [reply ledger/ledger-security-test-tool/BUILD.bazel:27]: You can still independent targets but in the same package (dir)

File: ledger/ledger-api-tests/suites/BUILD.bazel [hunk 4]
```diff
@@ -41,6 +41,10 @@ versions = {
     da_scala_library(
         name = "suites-%s" % lf_version,
         srcs = config["entrypoints"] + config["srcs"],
+        scala_deps = [
+            "@maven//:org_scalatest_scalatest_core",
```
Comment (tudor-da) [inline ledger/ledger-api-tests/suites/BUILD.bazel:45]: This looks odd. In the final implementation, I'd decouple the tagging annotations from the Scalatest lib, at least for the scope of the ledger-api-test-tool. Fine for the PoC scope, though

File: ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/TransactionServiceAuthorizationIT.scala [hunk 5]
```diff
@@ -83,6 +95,14 @@ class TransactionServiceAuthorizationIT extends LedgerTestSuite {
     "TXRejectMultiActorMissingAuth",
     "Reject exercising a multi-actor choice with missing authorizers",
     allocate(TwoParties, SingleParty),
+    tags = List(
```
Comment (tudor-da) [inline ledger/ledger-api-tests/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_8/TransactionServiceAuthorizationIT.scala:98]: I wonder whether we can avoid injecting the security evidencing tags as arguments and instead add them as annotations on the tests methods. I think the separation of concerns here would make it clear that these are not affecting the scope of the test (like the other arguments such)

File: ledger/ledger-security-evidence/tool/src/scala/org/scalatest/daml/SystematicTestingGenerator.scala [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+      }
+    }
+  }
+
+  private def loadIntelliJClasspath(): Option[String] =
+    Some(System.getProperty("java.class.path")).filter(!_.contains("sbt-launch.jar"))
+
+  def main(args: Array[String]): Unit = {
+
+    val ledgerApiTests = (suites.v1_8.default(0L) ++ suites.v1_8.optional()).toList
```
Comment (tudor-da) [inline ledger/ledger-security-evidence/tool/src/scala/org/scalatest/daml/SystematicTestingGenerator.scala:121]: Not ideal that we need to depend on the contents of `//ledger/ledger-api-tests/suites`. Can we do without? (e.g. see my other comment re annotations)

File: ledger/ledger-security-evidence/tool/src/scala/org/scalatest/daml/SystematicTestingGenerator.scala [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    println("Writing security tests inventory..")
+    File("/Users/sergeykisel/git/daml/security-tests.json").write(
+      scalaTestEntries[SecurityTest, SecurityTestSuite, SecurityTestEntry](
+        testSuites,
+        SecurityTestEntry,
+      ).asJson.spaces2
+    )
+
+    println("Writing reliability tests inventory..")
+    File("/Users/sergeykisel/git/daml/reliability-tests.json").write(
```
Comment (tudor-da) [inline ledger/ledger-security-evidence/tool/src/scala/org/scalatest/daml/SystematicTestingGenerator.scala:153]: Can you make it runnable for everyone?

File: ledger/ledger-security-evidence/tool/src/scala/org/scalatest/daml/SystematicTestingGenerator.scala (hunk 6 above)
Comment (skisel-da) [reply ledger/ledger-security-evidence/tool/src/scala/org/scalatest/daml/SystematicTestingGenerator.scala:121]: How would that fix the situation? We would still need an access to these on the classpath level

File: ledger/ledger-security-evidence/tool/src/scala/org/scalatest/daml/SystematicTestingGenerator.scala (hunk 7 above)
Comment (skisel-da) [reply ledger/ledger-security-evidence/tool/src/scala/org/scalatest/daml/SystematicTestingGenerator.scala:153]: sure, leftover

File: ledger/ledger-api-tests/suites/BUILD.bazel (hunk 4 above)
Comment (skisel-da) [reply ledger/ledger-api-tests/suites/BUILD.bazel:45]: I've added this concern to the design doc. The lib part of the approach would not have scalatest dependency and only contain tags. (In the form of annotations or case classes, to decide later)

**Pull request closed without merging**